static/src/javascripts/lib/cmp.spec.js

```diff
@@ -0,0 +1,215 @@
+// @flow
+import { init, onConsentNotification, consentState, _ } from 'lib/cmp';
+import { getAdConsentState as _getAdConsentState } from 'common/modules/commercial/ad-prefs.lib';
+
+const getAdConsentState: any = _getAdConsentState;
+
+jest.mock('common/modules/commercial/ad-prefs.lib', () => ({
+    getAdConsentState: jest.fn(),
+    thirdPartyTrackingAdConsent: jest.fn(),
+}));
+
+describe('cmp', () => {
+    beforeEach(() => {
+        _.resetCmp();
+        getAdConsentState.mockReset();
+    });
+
+    it('does not re-trigger consent notifications if init called multiple times', () => {
+        const myCallBack = jest.fn();
+
+        init();
+
+        onConsentNotification('functional', myCallBack);
+
+        expect(myCallBack).toHaveBeenCalledTimes(1);
+
+        init();
+
+        expect(myCallBack).toHaveBeenCalledTimes(1);
+    });
+
+    describe('consentState', () => {
+        it('returns null state if cmp lib has not been initialised', () => {
+            expect(consentState('functional')).toBeNull();
+            expect(consentState('performance')).toBeNull();
+            expect(consentState('advertisement')).toBeNull();
+        });
+
+        it('returns functional consent state', () => {
```

Comment:
I think this test (and all the similar ones) is a bit misleading.

The test says "returns functional consent state", but is actually just testing the initial state for each category. I don't think "initial state of functional == true" is a test that we really want to be seeing, since that's explicitly not correct.

These tests need a way of injecting a cookie value into the state so we can test the following realistic application states:
* a cookie does not exist
* a cookie exists with consent given
* a cookie exists with consent explicitly denied

Reply:
> The test says "returns functional consent state", but is actually just testing the initial state for each category. 

This statement is correct, as we're setting `purposes.functional.state = true;` as the initial state right now as discussed elsewhere and the tests are doing nothing to alter that initial state.

> I don't think "initial state of functional == true" is a test that we really want to be seeing, since that's explicitly not correct.

As mentioned in my comment https://github.com/guardian/frontend/pull/21696#discussion_r310176331 should we introduce a switch which states whether the new CMP UI is ON/OFF? If we do this we can test against the initial state being `true` when it's OFF and `null` when it's ON, this would be until we introduce the logic to read the cookie, at which point we can mock different values for the cookie to assert against?

Reply:
We could certainly adds some tests which simulate a change of state to be sure that `consentState` is always returning the latest state, because as you correct pointed out we're only checking whether it returns the default state right now.

Reply:
We've had a chat about this IRL and I now understand the problem being solved here. Let's pick it up in a bit more detail, and I think if we split this PR up into some more focused pieces we'll be able to more easily understand the implications of these settings.